src/main/java/meteordevelopment/meteorclient/utils/misc/PooledList.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * This file is part of the Meteor Client distribution (https://github.com/MeteorDevelopment/meteor-client).
+ * Copyright (c) Meteor Development.
+ */
+
+package meteordevelopment.meteorclient.utils.misc;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class PooledList<T> implements Iterable<T> {
+    protected final Pool<T> pool;
+    protected final List<T> list = new ArrayList<>();
+
+    public PooledList(Producer<T> producer) {
+        pool = new Pool<>(producer);
+    }
+
+    public T get() {
+        T obj = pool.get();
+        list.add(obj);
+        return obj;
+    }
+
+    public T peekList() {
+        return list.get(0);
+    }
+
+    public void clear() {
+        for (Iterator<T> it = list.iterator(); it.hasNext(); ) {
+            pool.free(it.next());
+            it.remove();
+        }
+        for (T obj : list) free(obj);
+        list.clear();
+    }
+
+    public void free(T obj) {
+        pool.free(obj);
+        list.remove(obj);
+    }
+
+    /** Required instead of {@link PooledList#free(Object)} when in an iterator loop. {@link PooledList} does not support freeing in foreach loops. */
+    public void free(T obj, Iterator<T> iterator) {
```

Comment:
Instead of having this method what you can do is make a custom `private class PooledListIterator implements Iterator<T> { ... }` which will basically be a wrapper over the normal iterator returned from `list.iterator()` but it will override the `remove()` method to call `super.remove();` and `free(obj);`. Then in the `iterator()` method you can just return a new instance of the custom iterator and code using this class can just normally call `it.remove()`.

Reply:
I genuinely have no clue how to implement it, currently it uses `list.iterator()` which is a private implementation that I can’t just extend so I’d have to reimplement it or something